src/pubsub/ua_pubsub_keystorage.c

```diff
@@ -0,0 +1,530 @@
+    if(!keyStorage)
+        return NULL;
+
+    /*initialize it to have the possibility to delete it in case of error somewhere*/
+    memset(keyStorage, 0, sizeof(UA_PubSubKeyStorage));
+
+    return keyStorage;
+}
+
+UA_PubSubKeyStorage *
+UA_PubSubKeyStorage_getKeyStorageWithSecurityGroupId(UA_Server *server,
+                                                     UA_String securityGroupId) {
+
+    if(!server || UA_String_isEmpty(&securityGroupId))
+        return NULL;
+
+    UA_PubSubKeyStorage *outKeyStorage, *keyStorageTemp;
+    LIST_FOREACH_SAFE(outKeyStorage, &server->pubSubManager.pubSubKeyList, keyStorageList,
+                      keyStorageTemp) {
+        if(UA_String_equal(&outKeyStorage->securityGroupID, &securityGroupId)) {
+            return outKeyStorage;
+        }
+    }
+    return NULL;
+}
+
+UA_StatusCode
+UA_PubSubKeyStorage_getPubSubSecurityPolicyByUri(UA_Server *server,
+                                                 const UA_String *securityPolicyUri,
+                                                 UA_PubSubSecurityPolicy **policy) {
+    if(!server || !securityPolicyUri)
+        return UA_STATUSCODE_BADINVALIDARGUMENT;
+
+    UA_StatusCode retval = UA_STATUSCODE_BAD;
+
+    UA_ServerConfig *config = UA_Server_getConfig(server);
+
+    for(size_t i = 0; i < config->pubSubConfig.securityPoliciesSize; i++) {
+        if(UA_String_equal(securityPolicyUri,
+                           &config->pubSubConfig.securityPolicies[i].policyUri)) {
+            *policy = &config->pubSubConfig.securityPolicies[i];
+            retval = UA_STATUSCODE_GOOD;
+            break;
+        }
+    }
+    return retval;
+}
+
+static void
+UA_PubSubKeyStorage_clearKeyList(UA_PubSubKeyStorage *keyStorage) {
+    if(keyStorage->keyList) {
+        UA_PubSubKeyListItem *iterator = keyStorage->firstItem;
+        UA_PubSubKeyListItem *head = iterator;
+        while(iterator != NULL) {
+            if(iterator->key.data) {
+                UA_free(iterator->key.data);
+                head = iterator;
+                iterator = iterator->next;
+                UA_free(head);
+            } else
+                break;
+        }
+        keyStorage->keyListSize = 0;
+    }
+}
+
+void
+UA_PubSubKeyStorage_deleteMembers(UA_Server *server, UA_PubSubKeyStorage *keyStorage) {
+
+    if(!keyStorage)
+        return;
+
+    /*Remove callback*/
+    if(server) {
+        if(!keyStorage->callBackId) {
+            removeCallback(server, keyStorage->callBackId);
+            keyStorage->callBackId = 0;
+        }
+    }
+
+    UA_PubSubKeyStorage_clearKeyList(keyStorage);
+
+    if(keyStorage->securityGroupID.data)
+        UA_String_clear(&keyStorage->securityGroupID);
+
+    memset(keyStorage, 0, sizeof(UA_PubSubKeyStorage));
+}
+
+void
+UA_PubSubKeyStorage_delete(UA_Server *server, UA_PubSubKeyStorage *keyStorage) {
+
+    if(!keyStorage)
+        return;
+
+    UA_PubSubKeyStorage_deleteMembers(server, keyStorage);
```

Comment:
If this is the only place that uses UA_PubSubKeyStorage_deleteMembers, then the indirection can be removed.
Just define UA_PubSubKeyStorage_delete in that case.

Reply:
deleteMembers is now changed to clear and delete is kept because clear does not free the keystorage pointer.